realm/realm-library/src/main/java/io/realm/HandlerController.java

```diff
@@ -257,58 +261,66 @@ private void updateAsyncEmptyRealmObject() {
+        // TODO: Why is this here? This was not called in `completedAsyncQueriesUpdate()`. Problem?
         if (!realm.isClosed() && threadContainsAsyncEmptyRealmObject()) {
             updateAsyncEmptyRealmObject();
         }
-        // It is very important to notify the global listeners last.
-        // We don't sync RealmResults in realmChanged, instead, they are synced in notifySyncRealmResultsCallbacks.
-        // This is because of we need to compare the TableView version in order to decide if it changes. Thus, we cannot
-        // sync the RealmResults together with advance read - the result's listener won't get called.
-        // NotificationTest.callingOrdersOfListeners will fail if orders change.
+
+        // Trigger global listeners last.
+        // Note that NotificationTest.callingOrdersOfListeners will fail if orders change.
         notifyGlobalListeners();
     }
 
-    private void notifyTypeBasedListeners() {
-        notifyAsyncRealmResultsCallbacks();
-        notifySyncRealmResultsCallbacks();
-        notifyRealmObjectCallbacks();
+    private void collectAsyncRealmResultsCallbacks(List<RealmResults<? extends RealmModel>> resultsToBeNotified) {
+        collectRealmResultsCallbacks(asyncRealmResults.keySet().iterator(), resultsToBeNotified);
     }
 
-    private void notifyAsyncRealmResultsCallbacks() {
-        notifyRealmResultsCallbacks(asyncRealmResults.keySet().iterator());
+    private void collectSyncRealmResultsCallbacks(List<RealmResults<? extends RealmModel>> resultsToBeNotified) {
+        collectRealmResultsCallbacks(syncRealmResults.keySet().iterator(), resultsToBeNotified);
     }
 
-    private void notifySyncRealmResultsCallbacks() {
-        notifyRealmResultsCallbacks(syncRealmResults.keySet().iterator());
-    }
 
-    private void notifyRealmResultsCallbacks(Iterator<WeakReference<RealmResults<? extends RealmModel>>> iterator) {
-        List<RealmResults<? extends RealmModel>> resultsToBeNotified =
-                new ArrayList<RealmResults<? extends RealmModel>>();
+    private void collectRealmResultsCallbacks(Iterator<WeakReference<RealmResults<? extends RealmModel>>> iterator,
+                                              List<RealmResults<? extends RealmModel>> resultsToBeNotified) {
         while (iterator.hasNext()) {
             WeakReference<RealmResults<? extends RealmModel>> weakRealmResults = iterator.next();
             RealmResults<? extends RealmModel> realmResults = weakRealmResults.get();
             if (realmResults == null) {
                 iterator.remove();
             } else {
-                // It should be legal to modify asyncRealmResults and syncRealmResults in the listener
+                // Sync the RealmResult so it is completely up to date.
+                // This is a prerequisite to calling the listener, so when the listener is finally triggered, all
+                // RealmResults will be up to date.
+                realmResults.syncIfNeeded();
```

Comment:
Would it be possible to separate this step after collecting all listeners especially in `realmChanged()` that it would look like             
```java
collectAsyncRealmResultsCallbacks(resultsToBeNotified);
collectSyncRealmResultsCallbacks(resultsToBeNotified); 
// sync if needed
enforceSyncCollectedRealmResultsCallbaks(resultsToBeNotified);
```

It might take another loop, but separating this step might help others not to forget and reuse.